File: pathplannerlib/src/test/native/cpp/pathplanner/lib/PathPointTest.cpp
#include <gtest/gtest.h>
#include "pathplanner/lib/PathPoint.h"

using namespace pathplanner;

#define DELTA 1E-2

TEST(PathPointTest, TestConstructor) {
	PathPoint p(frc::Translation2d(1.2_m, 2.7_m), frc::Rotation2d(25_deg), frc::Rotation2d(67_deg), 2.4_mps);

	EXPECT_EQ(frc::Translation2d(1.2_m, 2.7_m), p.m_position);
	EXPECT_EQ(frc::Rotation2d(25_deg), p.m_heading);
	EXPECT_EQ(frc::Rotation2d(67_deg), p.m_holonomicRotation);
	EXPECT_EQ(2.4_mps, p.m_velocityOverride);
}

TEST(PathPointTest, TestFromCurrentHolonomicState) {
	frc::Pose2d pose(1.7_m, 2.1_m, frc::Rotation2d(45_deg));
	frc::ChassisSpeeds speeds = frc::ChassisSpeeds::FromFieldRelativeSpeeds(1.7_mps, -1.2_mps, 0.8_rad_per_s, 0_deg);
	PathPoint p = PathPoint::fromCurrentHolonomicState(pose, speeds);

	EXPECT_EQ(frc::Translation2d(1.7_m, 2.1_m), p.m_position);
	EXPECT_NEAR(-35.22, p.m_heading.Degrees()(), DELTA);
	EXPECT_NEAR(45.0, p.m_holonomicRotation.Degrees()(), DELTA);
	EXPECT_NEAR(2.08, p.m_velocityOverride(), DELTA);
}

TEST(PathPointTest, TestFromCurrentDifferentialState) {
	frc::Pose2d pose(1.7_m, 2.1_m, frc::Rotation2d(45_deg));
	frc::ChassisSpeeds speeds = frc::ChassisSpeeds::FromFieldRelativeSpeeds(1.7_mps, 0_mps, 0.8_rad_per_s, 0_deg);
	PathPoint p = PathPoint::fromCurrentDifferentialState(pose, speeds);

	EXPECT_EQ(frc::Translation2d(1.7_m, 2.1_m), p.m_position);
	EXPECT_NEAR(45.0, p.m_heading.Degrees()(), DELTA);
	EXPECT_NEAR(0.0, p.m_holonomicRotation.Degrees()(), DELTA);
	EXPECT_NEAR(1.7, p.m_velocityOverride(), DELTA);
}

File: pathplannerlib/src/test/java/com/pathplanner/lib/PathPointTest.java
package com.pathplanner.lib;

import static org.junit.jupiter.api.Assertions.assertEquals;

import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import org.junit.jupiter.api.Test;

public class PathPointTest {
  public static final double DELTA = 1e-2;

  @Test
  public void constructor() {
    PathPoint p =
        new PathPoint(
            new Translation2d(1.2, 2.7),
            Rotation2d.fromDegrees(25),
            Rotation2d.fromDegrees(67),
            2.4);

    assertEquals(new Translation2d(1.2, 2.7), p.position);
    assertEquals(Rotation2d.fromDegrees(25), p.heading);
    assertEquals(Rotation2d.fromDegrees(67), p.holonomicRotation);
    assertEquals(2.4, p.velocityOverride, DELTA);
  }

  @Test
  public void fromCurrentHolonomicState() {
    Pose2d pose = new Pose2d(1.7, 2.1, Rotation2d.fromDegrees(45));
    ChassisSpeeds speeds = new ChassisSpeeds(1.7, -1.2, 0.8);

    PathPoint p = PathPoint.fromCurrentHolonomicState(pose, speeds);
    assertEquals(new Translation2d(1.7, 2.1), p.position);
    assertEquals(-35.22, p.heading.getDegrees(), DELTA);
    assertEquals(45, p.holonomicRotation.getDegrees(), DELTA);
    assertEquals(2.08, p.velocityOverride, DELTA);
  }

  @Test
  public void fromCurrentDifferentialState() {
    Pose2d pose = new Pose2d(1.7, 2.1, Rotation2d.fromDegrees(45));
    ChassisSpeeds speeds = new ChassisSpeeds(1.7, 0.0, 0.8);

    PathPoint p = PathPoint.fromCurrentDifferentialState(pose, speeds);
    assertEquals(new Translation2d(1.7, 2.1), p.position);
    assertEquals(45.0, p.heading.getDegrees(), DELTA);
    assertEquals(0.0, p.holonomicRotation.getDegrees(), DELTA);
    assertEquals(1.7, p.velocityOverride, DELTA);
  }
}
